pkg/providers/openstack/stage.go

```diff
@@ -0,0 +1,201 @@
+	return image, nil
+}
+
+// pickFlavor picks flavor that best matches criteria (i.e. HDD size and RAM size).
+// While diskMB is required, memoryMB is optional (set to -1 to ignore).
+func pickFlavor(clientNova *gophercloud.ServiceClient, diskMB int, memoryMB int) (*flavors.Flavor, error) {
+	if diskMB <= 0 {
+		return nil, errors.New("Please specify disk size.", nil)
+	}
+
+	var flavs []flavors.Flavor = listFlavors(clientNova, int(math.Ceil(float64(diskMB)/1024)), memoryMB)
+
+	// Find smallest flavor for given conditions.
+	logrus.Infof("Find smallest flavor for conditions: diskMB >= %d AND memoryMB >= %d\n", diskMB, memoryMB)
+
+	var bestFlavor flavors.Flavor
+	var minDiffDisk int = -1
+	var minDiffMem int = -1
+	for _, f := range flavs {
+		diffDisk := f.Disk*1024 - diskMB
+		var diffMem int = 0 // 0 is best value
+		if memoryMB > 0 {
+			diffMem = f.RAM - memoryMB
+		}
+
+		if diffDisk >= 0 && // disk is big enough
+			(minDiffDisk == -1 || minDiffDisk > diffDisk) && // disk is smaller than current best, but still big enough
+			diffMem >= 0 && // memory is big enough
+			(minDiffMem == -1 || minDiffMem >= diffMem) { // memory is smaller than current best, but still big enough
+			bestFlavor, minDiffDisk, minDiffMem = f, diffDisk, diffMem
+		}
+	}
+	if minDiffDisk == -1 {
+		return nil, errors.New(fmt.Sprintf("No flavor fits required conditions: diskMB >= %d AND memoryMB >= %d\n", diskMB, memoryMB), nil)
+	}
+	return &bestFlavor, nil
+}
+
+// listFlavors returns list of all flavors.
+func listFlavors(clientNova *gophercloud.ServiceClient, minDiskGB int, minMemoryMB int) []flavors.Flavor {
+	var flavs []flavors.Flavor = make([]flavors.Flavor, 0)
+
+	pagerFlavors := flavors.ListDetail(clientNova, flavors.ListOpts{
+		MinDisk: minDiskGB,
+		MinRAM:  minMemoryMB,
+	})
+	pagerFlavors.EachPage(func(page pagination.Page) (bool, error) {
+		flavorList, _ := flavors.ExtractFlavors(page)
+
+		for _, f := range flavorList {
+			flavs = append(flavs, f)
+		}
+
+		return true, nil
+	})
+	return flavs
+}
+
+// pushImage first creates meta for image at OpenStack, then it sends binary data for it, the qcow2 image.
+func pushImage(clientGlance *gophercloud.ServiceClient, imageName string, imageFilepath string, flavor *flavors.Flavor) (*images.Image, error) {
+	// Create metadata (on OpenStack).
+	createdImage, err := createImage(clientGlance, imageName, flavor)
+	if err != nil {
+		return nil, err
+	}
+
+	// Send the image binary data to OpenStack
+	if err = uploadImage(clientGlance, createdImage.ID, imageFilepath); err != nil {
+		return nil, err
+	}
+
+	return createdImage, nil
+}
+
+// createImage creates image metadata on OpenStack.
+func createImage(clientGlance *gophercloud.ServiceClient, name string, flavor *flavors.Flavor) (*images.Image, error) {
+	createdImage, err := images.Create(clientGlance, images.CreateOpts{
+		Name:             name,
+		Tags:             []string{"tagOSv", "tagCapstan"},
```

Comment:
this shouldn't be hardcoded. it doesn't look like it's used anywhere else; maybe just delete it?